File: core/serial_manager.py
"""
串口通信管理模块 (统一协议版本)
支持新的高效协议和旧格式兼容
"""
import serial
import serial.tools.list_ports
from PyQt6.QtCore import QObject, QThread, pyqtSignal, QMutex
from typing import Optional
import time

from core.unified_data_protocol import (
    SerialProtocolParser, UnifiedData, HandshakeInfo,
    StateDefinition, StateValue
)


class SerialWorker(QThread):
    """串口数据读取工作线程"""
    data_received = pyqtSignal(object)      # UnifiedData
    handshake_received = pyqtSignal(object)  # HandshakeInfo
    error_occurred = pyqtSignal(str)
    connection_lost = pyqtSignal()
    
    def __init__(self, port: str, baudrate: int):
        super().__init__()
        self.port = port
        self.baudrate = baudrate
        self._running = False
        self._serial: Optional[serial.Serial] = None
        self._mutex = QMutex()
        self._start_time = 0.0
        self._parser = SerialProtocolParser()

    def run(self):
        """线程主循环"""
        try:
            self._serial = serial.Serial(
                port=self.port,
                baudrate=self.baudrate,
                timeout=0.1,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE
            )
            self._running = True
            self._start_time = time.time()
            self._parser.reset()

            buffer = ""
            while self._running:
                if self._serial and self._serial.in_waiting:
                    try:
                        # 读取可用数据
                        chunk = self._serial.read(self._serial.in_waiting)
                        buffer += chunk.decode('utf-8', errors='ignore')
                        
                        # 按行处理
                        while '\n' in buffer:
                            line, buffer = buffer.split('\n', 1)
                            line = line.strip()
                            if line:
                                self._process_line(line)

                    except serial.SerialException as e:
                        self.error_occurred.emit(f"读取错误: {str(e)}")
                        self.connection_lost.emit()
                        break
                else:
                    self.msleep(1)

        except serial.SerialException as e:
            self.error_occurred.emit(f"串口打开失败: {str(e)}")
        finally:
            self._close_serial()
    
    def _process_line(self, line: str):
        """处理一行数据"""
        frame_type, data = self._parser.parse_line(line)

        if frame_type == 'handshake' and data:
            self.handshake_received.emit(data)
        elif frame_type == 'data' and data:
            # 如果是旧格式数据，补充时间戳
            if data.timestamp == 0.0:
                data.timestamp = time.time() - self._start_time
            self.data_received.emit(data)

            # 如果是首次数据且解析器自动生成了握手信息
            if self._parser.handshake and not hasattr(self, '_handshake_sent'):
                self._handshake_sent = True
                self.handshake_received.emit(self._parser.handshake)

    def _close_serial(self):
        """关闭串口"""
        if self._serial and self._serial.is_open:
            try:
                self._serial.close()
            except:
                pass
        self._serial = None
    
    def stop(self):
        """停止线程"""
        self._running = False
        self.wait(2000)
        self._close_serial()
    
    def send_data(self, data: str):
        """发送数据到串口"""
        self._mutex.lock()
        try:
            if self._serial and self._serial.is_open:
                self._serial.write((data + '\n').encode('utf-8'))
                self._serial.flush()
        except serial.SerialException as e:
            self.error_occurred.emit(f"发送失败: {str(e)}")
        finally:
            self._mutex.unlock()


class SerialManager(QObject):
    """串口管理器"""
    data_received = pyqtSignal(object)       # UnifiedData
    handshake_received = pyqtSignal(object)  # HandshakeInfo
    connection_changed = pyqtSignal(bool)    # True=已连接, False=已断开
    error_occurred = pyqtSignal(str)
    
    def __init__(self):
        super().__init__()
        self._worker: Optional[SerialWorker] = None
        self._is_connected = False
        self._handshake: Optional[HandshakeInfo] = None

    @staticmethod
    def get_available_ports() -> list[tuple[str, str]]:
        """获取可用串口列表"""
        ports = []
        for port in serial.tools.list_ports.comports():
            ports.append((port.device, f"{port.device} - {port.description}"))
        return ports
    
    @property
    def is_connected(self) -> bool:
        return self._is_connected
    
    @property
    def handshake_info(self) -> Optional[HandshakeInfo]:
        return self._handshake

    def connect(self, port: str, baudrate: int) -> bool:
        """连接串口"""
        if self._is_connected:
            self.disconnect()
        
        self._worker = SerialWorker(port, baudrate)
        self._worker.data_received.connect(self._on_data_received)
        self._worker.handshake_received.connect(self._on_handshake_received)
        self._worker.error_occurred.connect(self._on_error)
        self._worker.connection_lost.connect(self._on_connection_lost)
        self._worker.start()
        
        self._is_connected = True
        self.connection_changed.emit(True)
        return True

    def disconnect(self):
        """断开串口"""
        if self._worker:
            self._worker.stop()
            self._worker = None
        self._is_connected = False
        self._handshake = None
        self.connection_changed.emit(False)
    
    def send(self, data: str):
        """发送数据"""
        if self._worker:
            self._worker.send_data(data)
    
    def _on_data_received(self, data: UnifiedData):
        self.data_received.emit(data)
    
    def _on_handshake_received(self, handshake: HandshakeInfo):
        self._handshake = handshake
        self.handshake_received.emit(handshake)

    def _on_error(self, message: str):
        self.error_occurred.emit(message)
    
    def _on_connection_lost(self):
        self._is_connected = False
        self._handshake = None
        self.connection_changed.emit(False)

File: src/core/SerialManager.cpp
/**
 * SerialManager.cpp - 串口通信管理实现
 */

#include "core/SerialManager.h"

#include <QSerialPortInfo>
#include <QElapsedTimer>
#include <QDebug>

// ============ SerialWorker 实现 ============

SerialWorker::SerialWorker(const QString& port, int baudrate, QObject* parent)
    : QThread(parent)
    , m_port(port)
    , m_baudrate(baudrate)
{
}

SerialWorker::~SerialWorker() {
    stop();
}

void SerialWorker::run() {
    QElapsedTimer timer;
    timer.start();

    try {
        m_serial = new QSerialPort();
        m_serial->setPortName(m_port);
        m_serial->setBaudRate(m_baudrate);
        m_serial->setDataBits(QSerialPort::Data8);
        m_serial->setParity(QSerialPort::NoParity);
        m_serial->setStopBits(QSerialPort::OneStop);

        if (!m_serial->open(QIODevice::ReadWrite)) {
            emit errorOccurred(QString("串口打开失败: %1").arg(m_serial->errorString()));
            closeSerial();
            return;
        }

        m_running = true;
        m_startTime = timer.elapsed() / 1000.0;
        m_parser.reset();

        QString buffer;

        while (m_running) {
            if (m_serial->waitForReadyRead(10)) {
                QByteArray data = m_serial->readAll();
                buffer += QString::fromUtf8(data);

                // 按行处理
                while (buffer.contains('\n')) {
                    int idx = buffer.indexOf('\n');
                    QString line = buffer.left(idx).trimmed();
                    buffer = buffer.mid(idx + 1);

                    if (!line.isEmpty()) {
                        processLine(line);
                    }
                }
            }
        }
    } catch (const std::exception& e) {
        emit errorOccurred(QString("串口错误: %1").arg(e.what()));
    }

    closeSerial();
}

void SerialWorker::processLine(const QString& line) {
    auto [frameType, data] = m_parser.parseLine(line);

    if (frameType == "handshake" && data.has_value()) {
        HandshakeInfo info = data->value<HandshakeInfo>();
        emit handshakeReceived(info);
    } else if (frameType == "data" && data.has_value()) {
        UnifiedData udata = data->value<UnifiedData>();

        // 如果是旧格式数据，补充时间戳
        if (udata.timestamp == 0.0) {
            QElapsedTimer timer;
            udata.timestamp = timer.elapsed() / 1000.0 - m_startTime;
        }

        emit dataReceived(udata);

        // 首次数据且自动生成了握手信息
        if (m_parser.handshake() && !m_handshakeSent) {
            m_handshakeSent = true;
            emit handshakeReceived(*m_parser.handshake());
        }
    }
}

void SerialWorker::stop() {
    m_running = false;
    if (!wait(2000)) {
        terminate();
        wait();
    }
    closeSerial();
}

void SerialWorker::closeSerial() {
    if (m_serial) {
        if (m_serial->isOpen()) {
            m_serial->close();
        }
        delete m_serial;
        m_serial = nullptr;
    }
}

void SerialWorker::sendData(const QString& data) {
    QMutexLocker locker(&m_mutex);
    if (m_serial && m_serial->isOpen()) {
        m_serial->write((data + "\n").toUtf8());
        m_serial->flush();
    }
}

// ============ SerialManager 实现 ============

SerialManager::SerialManager(QObject* parent)
    : QObject(parent)
{
    // 注册元类型
    qRegisterMetaType<UnifiedData>("UnifiedData");
    qRegisterMetaType<HandshakeInfo>("HandshakeInfo");
}

SerialManager::~SerialManager() {
    disconnect();
}

QList<QPair<QString, QString>> SerialManager::getAvailablePorts() {
    QList<QPair<QString, QString>> ports;

    for (const QSerialPortInfo& info : QSerialPortInfo::availablePorts()) {
        QString description = QString("%1 - %2").arg(info.portName(), info.description());
        ports.append({info.portName(), description});
    }

    return ports;
}

bool SerialManager::connectToPort(const QString& port, int baudrate) {
    if (m_connected) {
        disconnect();
    }

    m_worker = std::make_unique<SerialWorker>(port, baudrate);

    connect(m_worker.get(), &SerialWorker::dataReceived,
            this, &SerialManager::onDataReceived);
    connect(m_worker.get(), &SerialWorker::handshakeReceived,
            this, &SerialManager::onHandshakeReceived);
    connect(m_worker.get(), &SerialWorker::errorOccurred,
            this, &SerialManager::onError);
    connect(m_worker.get(), &SerialWorker::connectionLost,
            this, &SerialManager::onConnectionLost);

    m_worker->start();
    m_connected = true;
    emit connectionChanged(true);

    return true;
}

void SerialManager::disconnect() {
    if (m_worker) {
        m_worker->stop();
        m_worker.reset();
    }
    m_connected = false;
    m_handshake.reset();
    emit connectionChanged(false);
}

void SerialManager::send(const QString& data) {
    if (m_worker) {
        m_worker->sendData(data);
    }
}

void SerialManager::onDataReceived(const UnifiedData& data) {
    emit dataReceived(data);
}

void SerialManager::onHandshakeReceived(const HandshakeInfo& info) {
    m_handshake = std::make_unique<HandshakeInfo>(info);
    emit handshakeReceived(info);
}

void SerialManager::onError(const QString& message) {
    emit errorOccurred(message);
}

void SerialManager::onConnectionLost() {
    m_connected = false;
    m_handshake.reset();
    emit connectionChanged(false);
}
